File: statemachine.go
package main

import "fmt"

type Status string

const (
	Pending   Status = "pending"
	Running   Status = "running"
	Succeeded Status = "succeeded"
	Failed    Status = "failed"
	Retrying  Status = "retrying"
)

type Event string

const (
	EventStart   Event = "start"
	EventSucceed Event = "succeed"
	EventFail    Event = "fail"
	EventRetry   Event = "retry"
)

var transitions = map[Status]map[Event]Status{
	Pending: {
		EventStart: Running 
	},
	Running: {
		EventSucceed: Succeeded,
		EventFail: Failed,
	},
	Failed: {
		EventRetry: Retrying, 
	},
	Retrying: {
		EventStart: Running,
	}
}

func transition(current Status, event Event) (Status,error) {
	events, ok := transitions[current]
	if !ok {
		return current, fmt.Errorf("no transitions from %q", current)
	}
	next, ok := events[event]
	if !ok {
		return current, fmt.Errorf("event %q invalid from %q", event,current)
	}

	return next,nil
}

File: src/anchora/state_machine.py
from __future__ import annotations

from enum import StrEnum


class Status(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    RETRYING = "retrying"


class Event(StrEnum):
    START = "start"
    SUCCEED = "succeed"
    FAIL = "fail"
    RETRY = "retry"


_TRANSITIONS: dict[Status, dict[Event, Status]] = {
    Status.PENDING: {
        Event.START: Status.RUNNING,
    },
    Status.RUNNING: {
        Event.SUCCEED: Status.SUCCEEDED,
        Event.FAIL: Status.FAILED,
    },
    Status.FAILED: {
        Event.RETRY: Status.RETRYING,
    },
    Status.RETRYING: {
        Event.START: Status.RUNNING,
    },
}


def transition(current: Status, event: Event) -> Status:
    events = _TRANSITIONS.get(current)
    if events is None:
        raise ValueError(f'no transitions from "{current}"')

    next_status = events.get(event)
    if next_status is None:
        raise ValueError(f'event "{event}" invalid from "{current}"')

    return next_status
